pkg/sql/sem/tree/constant.go

```diff
@@ -71,6 +71,12 @@ func typeCheckConstant(
 	if !desired.IsAmbiguous() {
 		for _, typ := range avail {
 			if desired.Equivalent(typ) {
+				if desired.UserDefined() && !desired.IsHydrated() {
+					desired, err = semaCtx.TypeResolver.ResolveTypeByOID(ctx, desired.Oid())
+					if err != nil {
+						return nil, err
+					}
+				}
```

Comment:
Is this fix required for CREATE INDEX? If so, then I feel like this commit should probably go somewhere before the last two commits? I was just thinking the last two commits should be "Enable CREATE INDEX by default" and "Disable CREATE INDEX in certain tests".

Reply:
Yes, let me re-order the commits this is needed for the partial index expressions, which are running into a bug where things don't get fully hydrated.

Reply:
I'm highly suspicious of this change. For a start, there's no guarantee that semaCtx.TypeResolver is not nil. Also, why isn't this new logic pushed down into the corresponding ResolveAsType implementation instead of being added here?